src/Data/Map/Internal.purs

```diff
@@ -158,6 +159,49 @@ instance traversableWithIndexMap :: TraversableWithIndex k (Map k) where
           <*> f k2 v2
           <*> traverseWithIndex f right
 
+-- | `SemigroupMap k v` provides a `Semigroup` instance for `Map k v` whose
+-- | definition depends on the `Semigroup` instance for the `v` type.
+-- | You should only use this type when you need `Data.Map` to have
+-- | a `Semigroup` instance.
+-- |
+-- | ```purescript
+-- | let
+-- |   s :: forall key value. key -> value -> SemigroupMap key value
+-- |   s k v = SemigroupMap (singleton k v)
+-- |
+-- | (s 1     "foo") <> (s 1     "bar") == (s 1  "foobar")
+-- | (s 1 (First 1)) <> (s 1 (First 2)) == (s 1 (First 1))
+-- | (s 1  (Last 1)) <> (s 1  (Last 2)) == (s 1  (Last 2))
+-- | ```
+newtype SemigroupMap k v = SemigroupMap (Map k v)
```

Comment:
I'd suggest moving the newtype to the non-`Internal` module, since no part of it requires access to the insides of `Map`, does it?

Reply:
Done in latest commit.